fv/etcd_restart_test.go

```diff
@@ -0,0 +1,174 @@
+	)
+
+	BeforeEach(func() {
+		felixes, etcd, client = containers.StartNNodeEtcdTopology(2, containers.DefaultTopologyOptions())
+
+		// Install a default profile that allows all ingress and egress, in the absence of any Policy.
+		defaultProfile := api.NewProfile()
+		defaultProfile.Name = "default"
+		defaultProfile.Spec.LabelsToApply = map[string]string{"default": ""}
+		defaultProfile.Spec.Egress = []api.Rule{{Action: api.Allow}}
+		defaultProfile.Spec.Ingress = []api.Rule{{Action: api.Allow}}
+		_, err := client.Profiles().Create(utils.Ctx, defaultProfile, utils.NoOptions)
+		Expect(err).NotTo(HaveOccurred())
+
+		// Wait until the tunl0 device appears; it is created when felix inserts the ipip module
+		// into the kernel.
+		Eventually(func() error {
+			links, err := netlink.LinkList()
+			if err != nil {
+				return err
+			}
+			for _, link := range links {
+				if link.Attrs().Name == "tunl0" {
+					return nil
+				}
+			}
+			return errors.New("tunl0 wasn't auto-created")
+		}).Should(BeNil())
+
+		// Create workloads, using that profile.  One on each "host".
+		for ii := range w {
+			wIP := fmt.Sprintf("10.65.%d.2", ii)
+			wIface := fmt.Sprintf("cali1%d", ii)
+			wName := fmt.Sprintf("w%d", ii)
+			w[ii] = workload.Run(felixes[ii], wName, wIface, wIP, "8055", "tcp")
+			w[ii].Configure(client)
+		}
+
+		cc = &workload.ConnectivityChecker{}
+	})
+
+	AfterEach(func() {
+		if CurrentGinkgoTestDescription().Failed {
+			for _, felix := range felixes {
+				felix.Exec("iptables-save", "-c")
+				felix.Exec("ipset", "list")
+				felix.Exec("ip", "r")
+			}
+		}
+
+		for _, wl := range w {
+			wl.Stop()
+		}
+		for _, felix := range felixes {
+			felix.Stop()
+		}
+
+		if CurrentGinkgoTestDescription().Failed {
+			etcd.Exec("etcdctl", "ls", "--recursive", "/")
+		}
+		etcd.Stop()
+	})
+
+	It("should detect and recoonnect after the etcd connection is black-holed", func() {
+		By("having initial workload to workload connectivity", func() {
+			cc.ExpectSome(w[0], w[1])
+			cc.ExpectSome(w[1], w[0])
+			cc.CheckConnectivity()
+		})
+
+		By("silently dropping etcd packets", func() {
+			// Normally, if a connection closes at either end, the other peer's traffic will get
+			// FIN or RST responses, which cleanly shut down the connection.  However, in order
+			// to test the GRPC-level keep-alives, we want to simulate a network or NAT change that
+			// starts to black-hole the TCP connection so that there are no responses of any kind.
+			var portRegexp = regexp.MustCompile(`sport=(\d+).*dport=2379`)
+			for _, felix := range felixes {
+				// Use conntrack to identify the source port that Felix is using.
+				out, err := felix.ExecOutput("conntrack", "-L")
+				Expect(err).NotTo(HaveOccurred())
+				logrus.WithField("output", out).WithError(err).Info("Conntrack entries")
+				found := false
+				for _, line := range strings.Split(out, "\n") {
+					matches := portRegexp.FindStringSubmatch(line)
+					if len(matches) == 0 {
```

Comment:
Looks like "< 2" would be a more relevant check, given that code below uses `matches[1]`.